src/flag_gems/ops/cat.py

```diff
@@ -39,30 +127,101 @@ def cat(
             raise RuntimeError(
                 f"Tensors must have same number of dimensions: got {len(inp0_shape)} and {len(s)}"
             )
-    # Same size check
     for tensor_idx, inp_shape in enumerate(inp_shapes):
         for idx, (common_length, length) in enumerate(zip(inp0_shape, inp_shape)):
-            if idx == dim:
-                continue
-            elif length != common_length:
+            if idx != dim and length != common_length:
                 raise RuntimeError(
                     f"Sizes of tensors must match except in dimension {dim}. "
                     f"Expected size {common_length} but got size {length} for tensor number "
                     f"{tensor_idx} in the list"
                 )
 
-    out_shape = list(inp0_shape)
-    out_shape[dim] = sum(s[dim] for s in inp_shapes)
-    out0 = torch.empty(out_shape, dtype=A[0].dtype, device=A[0].device)
-    out0_strides = out0.stride()
-    out0_offsets = list(
-        itertools.accumulate(
-            [s[dim] * out0_strides[dim] for s in inp_shapes[:-1]], initial=0
+    dtype, device = A[0].dtype, A[0].device
+    shapes = [t.shape for t in A]
+    cat_dim_sizes = [s[dim] for s in shapes]
+    out_shape = list(shapes[0])
+    out_shape[dim] = sum(cat_dim_sizes)
+    out = torch.empty(out_shape, dtype=dtype, device=device)
+
+    BLOCK = 1024
+    dim_offset = 0
+
+    i = 0
+    while i < len(A):
+        tensors_in_batch = A[i : i + 4]
+        num_tensors_in_batch = len(tensors_in_batch)
+
+        args = []
+        total_elements_list = []
+        current_dim_offset = dim_offset
+
+        for j in range(4):
+            if j < num_tensors_in_batch:
+                tensor = tensors_in_batch[j].contiguous()
```

Comment:
I recommend reusing the `flag_gems.contiguous` to ensure that the triton kernel runs, even when it is called explicitly.